File: Leetcode(209-1239)/416_partition_equal_subset_sum.cpp
class Solution {
public:
    
    bool subset_sum(vector<int>&nums,int sum)
    {
        int n=nums.size();
        
        bool dp[n+1][sum+1];
        
        for(int j=0;j<=sum;j++)
        {
            dp[0][j]=0;
        }
        
        for(int i=0;i<=n;i++)
        {
            dp[i][0]=1;
        }
        
        dp[0][0]=1;
        
        for(int i=1;i<=n;i++)
        {
            for(int j=1;j<=sum;j++)
            {
                if(nums[i-1] <= j)
                {
                    dp[i][j]=dp[i-1][j] || dp[i-1][j-nums[i-1]];
                }
                else
                {
                    dp[i][j]=dp[i-1][j];
                }
            }
        }
        return dp[n][sum];
    }
    
    
    bool canPartition(vector<int>& nums) {
        int sum=0;
        for(auto num : nums)
        {
            sum+=num;
        }
        
        if(sum%2!=0)
        {
            return 0;
        }
    return subset_sum(nums,sum/2);

    }
};

File: Leetcode(209-1239)/416_partition_equal_subset_sum.py
class Solution:
            
    def canPartition(self, nums: List[int]) -> bool:
        def subset_sum(nums,target):
                arr=[[False for _ in range(target+1)] for _ in range(len(nums)+1)]
            
                for i in range(len(nums)+1):
                    arr[i][0]=True
                for i in range(1,len(nums)+1):
                    for j in range(1,target+1):
                        if(j>=nums[i-1]):
                            arr[i][j]=arr[i-1][j-nums[i-1]] or arr[i-1][j]
                        else:
                            arr[i][j]=arr[i-1][j]
                return arr[len(nums)][target]

        arr_sum=sum(nums)
        if(arr_sum%2 == 1):
            return False
        else:
            return subset_sum(nums,int(arr_sum/2))